backend/src/v2/driver/driver.go

```diff
@@ -1188,43 +1214,235 @@ func resolveInputs(ctx context.Context, dag *metadata.DAG, iterationIndex *int,
+}
+
+// resolveUpstreamParametersConfig is just a config struct used to store the
+// input parameters of the resolveUpstreamParameters function.
+type resolveUpstreamParametersConfig struct {
+	ctx        context.Context
+	paramSpec  *pipelinespec.TaskInputsSpec_InputParameterSpec
+	dag        *metadata.DAG
+	pipeline   *metadata.Pipeline
+	mlmd       *metadata.Client
+	inputs     *pipelinespec.ExecutorInput_Inputs
+	name       string
+	paramError func(error) error
+}
+
+// resolveUpstreamParameters resolves input parameters that come from upstream
+// tasks. These tasks can be components/containers, which is relatively
+// straightforward, or DAGs, in which case, we need to traverse the graph until
+// we arrive at a component/container (since there can be n nested DAGs).
+func resolveUpstreamParameters(cfg resolveUpstreamParametersConfig) error {
+	taskOutput := cfg.paramSpec.GetTaskOutputParameter()
+	glog.V(4).Info("taskOutput: ", taskOutput)
+	if taskOutput.GetProducerTask() == "" {
+		return cfg.paramError(fmt.Errorf("producer task is empty"))
+	}
+	if taskOutput.GetOutputParameterKey() == "" {
+		return cfg.paramError(fmt.Errorf("output parameter key is empty"))
+	}
+	tasks, err := getDAGTasks(cfg.ctx, cfg.dag, cfg.pipeline, cfg.mlmd, nil)
+	if err != nil {
+		return cfg.paramError(err)
+	}
+
+	// The producer is the task that produces the output that we need to
+	// consume.
+	producer := tasks[taskOutput.GetProducerTask()]
+	glog.V(4).Info("producer: ", producer)
+	currentTask := producer
+	currentSubTaskMaybeDAG := true
+	// Continue looping until we reach a sub-task that is NOT a DAG.
+	for currentSubTaskMaybeDAG {
+		glog.V(4).Info("currentTask: ", currentTask.TaskName())
+		_, outputParametersCustomProperty, err := currentTask.GetParameters()
+		if err != nil {
+			return err
+		}
+		// If the current task is a DAG:
+		if *currentTask.GetExecution().Type == "system.DAGExecution" {
+			// Since currentTask is a DAG, we need to deserialize its
+			// output parameter map so that we can look its
+			// corresponding producer sub-task, reassign currentTask,
+			// and iterate through this loop again.
+			var outputParametersMap map[string]string
+			b, err := outputParametersCustomProperty["Output"].GetStructValue().MarshalJSON()
 			if err != nil {
-				return nil, artifactError(err)
+				return err
 			}
-			producer, ok := tasks[taskOutput.GetProducerTask()]
-			if !ok {
-				return nil, artifactError(fmt.Errorf("cannot find producer task %q", taskOutput.GetProducerTask()))
+			json.Unmarshal(b, &outputParametersMap)
+			glog.V(4).Info("Deserialized outputParametersMap: ", outputParametersMap)
+			subTaskName := outputParametersMap["producer_subtask"]
+			glog.V(4).Infof(
+				"Overriding currentTask, %v, output with currentTask's producer_subtask, %v, output.",
+				currentTask.TaskName(),
+				subTaskName,
+			)
+
+			// Reassign sub-task before running through the loop again.
+			currentTask = tasks[subTaskName]
```

Comment:
might need to handle some of these potential error cases, for example: 
*  if the producer task is not in outputParametersMap
* subtaskname is not in tasks